contracts/service-registry/src/contract/query.rs

```diff
@@ -41,10 +45,64 @@ pub fn active_verifiers(
         .into_report()?;
 
     if verifiers.len() < service.min_num_verifiers.into() {
-        Err(report!(ContractError::NotEnoughVerifiers))
-    } else {
-        Ok(verifiers)
+        return Err(report!(ContractError::NotEnoughVerifiers));
     }
+
+    Ok(match service.max_num_verifiers {
+        Some(max_verifiers) => select_top_verifiers(verifiers, max_verifiers, env.block.height),
+        _ => verifiers,
+    })
+}
+
+fn select_top_verifiers(
+    mut verifiers: Vec<WeightedVerifier>,
+    max_verifiers: u16,
+    block_height: u64,
+) -> Vec<WeightedVerifier> {
+    if verifiers.len() <= max_verifiers as usize {
+        return verifiers;
+    }
+
+    // Figure out the min weight that will make it into the set
+    let cutoff_weight = {
+        let nth = max_verifiers as usize - 1;
+        verifiers.select_nth_unstable_by(nth, |a, b| b.weight.cmp(&a.weight));
+        verifiers[nth].weight
+    };
+
+    // Now sort by weight descending, but for verifiers with the cutoff weight, break ties via pick_random
+    verifiers.sort_by(|a, b| {
+        b.weight.cmp(&a.weight).then_with(|| {
+            if a.weight == cutoff_weight {
+                pick_random(a, b, block_height)
+            } else {
+                Ordering::Equal
+            }
```

Comment:
are you aware that this is not a fair pick? It's just fuzzing the position in the initial sequence a bit

Reply:
How so? The algorithm basically sorts by weight and then within the sub slice of verifiers with the cutoff weight, assigns each a random number in the 256 bit range, and sorts by that. So the order of that sub slice should be random, no?

The pick_random function is a bit misnamed, it's not picking a random element each time it's called, but is deterministic based on block height and verifier address.